### PR TITLE
Reproducibility regardless of collection details

### LINKED ISSUE #500: Same seed, different result (with different way to collect data)
**Describe the bug**
The model result is deterministic, but in two different way. Different types of `adata` causes the results to be different.

**Minimal Working Example**
(This is a pluto notebook file, sorry for the extension, Github doesn't allow drag dropping `.jl` file). This is the example predator-prey model, I modified it so it uses seed and scheduler.

[agentsjl-collect-bug.txt](https://github.com/JuliaDynamics/Agents.jl/files/6560465/agentsjl-collect-bug.txt)

**Agents.jl version**
```julia
pkg> st Agents
     Project .........
      Status ............
  [46ada45e] Agents v4.2.4
```
**Additional infomation**

I ***think*** this happened because `init_agent_dataframe` uses two other functions, one of which uses `random_agent`, which uses `shuffle!(model.rng)` while the other doesn't. 

https://github.com/JuliaDynamics/Agents.jl/blob/c377d812146de06e30a13471b4b04f9076b8988e/src/simulations/collect.jl#L201-L206

https://github.com/JuliaDynamics/Agents.jl/blob/c377d812146de06e30a13471b4b04f9076b8988e/src/simulations/collect.jl#L211-L213

https://github.com/JuliaDynamics/Agents.jl/blob/c377d812146de06e30a13471b4b04f9076b8988e/src/core/model.jl#L190-L212

### PR DESCRIPTION
**Pull request by Libbum**
Fixes #500

@ndgnuh, can you confirm this fixes the issue from your point of view?

With your test case, I now get
```julia
adf1 = collect_agg(Schedulers.by_id)
adf2 = collect_normal(Schedulers.by_id)
adf1 == adf2 # True

adf1 = collect_agg(Schedulers.randomly)
adf2 = collect_normal(Schedulers.randomly)
adf1 == adf2 # True
```

### CODE REVIEW

**Comment by Libbum** [comment]
I didn't quite get what was happening with the video portion, so if that's still an issue after this fix, can you provide me a simple bit of code that shows the difference when using `abm_video`?

**Comment by ndgnuh** [comment]
I can confirm that the two `adf` is the same (since I have used the same approach in my fork and it works). For the video please wait so I come up with a smaller example. The video I show in the issue takes 10 minutes to render. I think it can be reproduced with the `randomly` scheduler.

**Comment by ndgnuh** [comment]
Sorry for the delay, I got a lot of stuffs going on, here's a simplified version of my model.

Basically, the data collection is ok now, even with the `randomly` scheduler.

image

Trying to replicate the simulation gives:
image

However, when generating the video with the same seed and the same scheduler, we got a totally different simulation, where every agents die midway.

https://user-images.githubusercontent.com/33273948/121125895-825d7080-c851-11eb-842e-ea9f90352994.mp4

The problem is not reproducible with schedulers like `by_id` (based on replication population image and video of seed 44)

image

https://user-images.githubusercontent.com/33273948/121126527-7a520080-c852-11eb-9276-c5b8f718b027.mp4



The source code for the whole experiment is here. Sorry for the extra package. Running the model should only takes 0.2s but making the video takes nearly 5m on my machine.  IDK how to optimize that though, maybe it's because I don't have a dedicated graphic card.
<details>
<summary>It's a Pluto notebook</summary>

```julia
### A Pluto.jl notebook ###
# v0.14.7

using Markdown
using InteractiveUtils

# ╔═╡ c79638a6-c80d-11eb-3254-b33f7a382f78
begin
	using Pkg
	Pkg.activate(joinpath(@__DIR__, ".."))
	using Random
	using Agents
	using DataFrames
	using ImageFiltering
	using CairoMakie
	using InteractiveDynamics
end

# ╔═╡ 7852fa86-a513-49ac-92b1-7f0a47cedc17
function neighbors_at(n::Integer)
    k = 2 * n + 1
    kern = zeros(Int, k, k)
    center = k ÷ 2 + 1
    dist = map(CartesianIndices(kern)) do I
        i, j = Tuple(I)
        center - (abs(i - center) + abs(j - center))
    end
    dist[center, center] = 0
    map(findall(dist .> 0)) do I
        (I[1] - center, I[2] - center)
    end
end

# ╔═╡ 6ed9031f-8754-46bf-bf92-304d1ea462b8
const default_parameters = (#
    energy_miss=0.025,
    age_init=168,
    age_reproduce=504,
    age_old=600,
    age_die=720,
    pr_reproduce=0.176,
    pr_egg_death=0.0025,
    pr_old_death=0.04,
    offspring_max=12,
    offspring_min=5,
    energy_max=1.0,
    energy_transfer=0.1,
    energy_consume=0.025,
    energy_move=0.2,
    energy_reproduce=0.8,
    move_directions=neighbors_at(2)
)

# ╔═╡ 55d77475-590d-4c21-a47b-2012fba13d8f
Base.@kwdef mutable struct BPH <: AbstractAgent
    id::Int
    pos::Dims{2}
    energy::Float16
    age::Int16
end

# ╔═╡ cc080461-80c3-40cd-bc6e-d1c46f017b9a
function model_step!(model)
	T = eltype(model.food)
    @. model.food = min(model.food + model.food * T(0.008), one(T))
end

# ╔═╡ c7c02074-b970-425a-9586-48aa34e62c27
function agent_step!(agent, model)
    # position
    x, y = agent.pos

    # Older
    agent.age = agent.age + one(agent.age)

    # Step wise energy loss
    agent.energy = agent.energy - (agent.age ≥ model.age_init) * model.energy_consume

    # Move conditionally
    if agent.age ≥ model.age_init && agent.energy ≥ model.energy_move
        walk!(agent, rand(model.rng, model.move_directions), model)
    end

    # Eat conditionally
    if model.food[x, y] > 0 && agent.age ≥ model.age_init
        transfer = min(#
            model.energy_transfer,
            model.food[x, y],
            model.energy_max - agent.energy,
        )
        model.food[x, y] -= transfer
        agent.energy += transfer
    end

    # Reproduce conditionally
    if (
        agent.age ≥ model.age_reproduce && # Old enough
        agent.energy ≥ model.energy_reproduce && # Energy requirement
        rand(model.rng) ≤ model.pr_reproduce # Have RNG Jesus by your side
    )
        nb_offspring = rand(model.rng, (model.offspring_min):(model.offspring_max))
        for _ in 1:nb_offspring
            id = nextid(model)
            agent = BPH(;
                id=id,
                pos=agent.pos,
                energy=0.4,
                age=0,
            )
            add_agent_pos!(agent, model)
        end
        agent.energy -= one(agent.energy) / 10
    end

    # Die conditionally
    if (agent.energy ≤ 0) || # Exausted
       (agent.age ≥ model.age_die) || # Too old
       (rand(model.rng) ≤ model.pr_killed[x, y]) || # killed
       (
           model.age_die > agent.age ≥ model.age_old && # Old
           rand(model.rng) ≤ model.pr_old_death # And weak
       ) ||
       (
           agent.age < model.age_init && # Young
           rand(model.rng) ≤ model.pr_egg_death # And weak
       ) # then
        kill_agent!(agent, model)
        return nothing
    end
    # End of agent step
end

# ╔═╡ a7a6cbab-5b38-42d3-9ede-1e2b5550f2f1
function init_model(; envmap, pr_killed, seed, scheduler, kwargs...)
	nb_bph_init = 200
    rng = MersenneTwister(seed)
    food = collect(transpose(envmap))
    pr_killed = imfilter(isnan.(food) * pr_killed, Kernel.gaussian(2.5))

    # PROPERTIES

    params = merge(default_parameters, kwargs)
    properties = (#
        food=food,
        total_bph=nb_bph_init,
        death_natural=0,
        death_predator=0,
        pr_killed=pr_killed,
        pr_killed_positions=convert.(Tuple, findall(!iszero, pr_killed)),
        energy_full=1.0 - params.energy_transfer,
        params...,
    )

    # MODEL

    space = GridSpace(size(food); periodic=false)
    model = ABM(BPH, space; scheduler=scheduler, properties=properties, rng=rng)

    # AGENTS CREATION
    positions = Iterators.product(1:5, 1:5)
    positions = filter(pos -> !isnan(food[pos...]), collect(positions))
    for _ in 1:nb_bph_init
        bph = BPH(; #
            id=nextid(model),
            pos=rand(model.rng, positions),
            energy=rand(model.rng, 0.4:0.01:0.6),
            age=rand(model.rng, 0:300),
        )
        add_agent_pos!(bph, model)
    end

    # RETURN
    return model
end

# ╔═╡ 79076b76-917f-439c-b954-8046867614c0
heatkwargs=(
	nan_color=RGBAf0(1.0, 1.0, 0.0, 0.5),
	colormap=[RGBAf0(0, 1.0, 0, i) for i in 0:0.01:1],
	colorrange=(0, 1),
)

# ╔═╡ 11b14ef0-3554-498c-8e14-aafd5d026a49
adata, mdata = let food(model) = count(model.food .≥ 0.5), 
	bph(agent) = agent.energy > 0
    [(bph, count)], [food]
end


# ╔═╡ 11362a40-572d-4019-9620-baecb02b01f9
begin
	function envmap(T::Type, d::Integer)
		env = ones(T, 100, 100)
		xstart, xend = (100 - d) ÷ 2, (100 + d) ÷ 2
		env[:, xstart+1:xend] .= convert(T, NaN)
		return env
	end
	envmap(d::Integer) = envmap(Float16, d)
end

# ╔═╡ 6cc30e3d-ca13-4ffb-b623-fcdb7c31936c
function run_and_collect_data(; seed, scheduler)
	model = init_model(; 
		envmap=envmap(20),
		pr_killed=0.075,
		seed=seed,
		scheduler=scheduler)
	adf, mdf = run!(model, agent_step!, model_step!, 2880; 
		adata=adata, mdata=mdata)
	leftjoin(adf, mdf, on=:step)
end

# ╔═╡ fe5717d3-dd4a-470a-82bb-453e9867dc3d
function run_and_collect_data2(; seed, scheduler)
	model = init_model(; 
		envmap=envmap(20),
		pr_killed=0.075,
		seed=seed,
		scheduler=scheduler)
	adata = [:energy]
	adf_, mdf = run!(model, agent_step!, model_step!, 2880; 
		adata=adata, mdata=mdata)
	adf = combine(groupby(adf_, :step), 
		:energy => (energy -> count(x -> x > 0, energy)) => :count_bph)
	leftjoin(adf, mdf, on=:step)
end

# ╔═╡ f1bbd543-4f24-496a-804e-c594ebc22c5c
function run_and_make_video(videopath; seed, scheduler)
	model = init_model(; 
		envmap=envmap(20),
		pr_killed=0.075,
		seed=seed,
		scheduler=scheduler)
	abm_video(videopath, model, agent_step!, model_step!;
		frames=2880,
		framerate=30,
		heatkwargs=heatkwargs,
		heatarray=model -> model.food,
		am=agent -> :circle
	)
	return videopath
end

# ╔═╡ dc557b3a-8f6d-4578-b723-60b8e02a02a7
let kwargs = (seed = rand(1:100), scheduler=Schedulers.randomly)
	df1=run_and_collect_data(;kwargs...)
	df2=run_and_collect_data2(;kwargs...)
	f = Figure()
	a1,a2 = Axis(f[1,1]), Axis(f[1,2])
	lines!(a1, df1.step, df1.count_bph)
	lines!(a1, df2.step, df2.count_bph)
	lines!(a2, df1.step, df1.food)
	lines!(a2, df2.step, df2.food)
	f
end

# ╔═╡ 8674a5e3-805a-4bfe-b8c1-1687a38df04e
let f = Figure()
	a = Axis(f[1,1])
	for seed in 1:20
		df = run_and_collect_data(;seed =seed, scheduler=Schedulers.randomly)
		kwargs = if df.food[end] ≤ df.food[1] ÷ 2
			lines!(a, df.step, df.count_bph; label = "Seed $seed")	
		else
			lines!(a, df.step, df.count_bph;)	
		end
	end
	f[1,2] = Legend(f, a, "Odd things", framevisible = false)
	f
end

# ╔═╡ 46abc4b7-6ed1-4adb-b31e-9973b06ef69d
run_and_make_video("Seed 4.mp4"; seed =4,scheduler = Schedulers.randomly)

# ╔═╡ Cell order:
# ╠═c79638a6-c80d-11eb-3254-b33f7a382f78
# ╠═7852fa86-a513-49ac-92b1-7f0a47cedc17
# ╠═6ed9031f-8754-46bf-bf92-304d1ea462b8
# ╠═55d77475-590d-4c21-a47b-2012fba13d8f
# ╠═cc080461-80c3-40cd-bc6e-d1c46f017b9a
# ╠═c7c02074-b970-425a-9586-48aa34e62c27
# ╠═a7a6cbab-5b38-42d3-9ede-1e2b5550f2f1
# ╠═79076b76-917f-439c-b954-8046867614c0
# ╠═11b14ef0-3554-498c-8e14-aafd5d026a49
# ╠═11362a40-572d-4019-9620-baecb02b01f9
# ╠═6cc30e3d-ca13-4ffb-b623-fcdb7c31936c
# ╠═fe5717d3-dd4a-470a-82bb-453e9867dc3d
# ╠═f1bbd543-4f24-496a-804e-c594ebc22c5c
# ╠═dc557b3a-8f6d-4578-b723-60b8e02a02a7
# ╠═8674a5e3-805a-4bfe-b8c1-1687a38df04e
# ╠═46abc4b7-6ed1-4adb-b31e-9973b06ef69d
```

</details>

**Comment by Libbum** [comment]
Thanks for the details @ndgnuh. Indeed, that looks odd still and I'm looking into it. 

The problem is most likely somewhere in InteractiveDynamics at this point, since we can now do
```julia
model = init_model(;
               envmap = envmap(20),
               pr_killed = 0.075,
               seed = seed,
               scheduler = scheduler,
           )
model2 = init_model(;
               envmap = envmap(20),
               pr_killed = 0.075,
               seed = seed,
               scheduler = scheduler,
           )
adf, mdf = run!(model, agent_step!, model_step!, 2880; adata, mdata)
adf2, mdf2 = run!(model2, agent_step!, model_step!, 2880; adata, mdata)
adf == adf2 #True
mdf == mdf2 #True
```

I don't see any clear issue in your plot function:
```julia
function run_and_make_video(videopath; seed, scheduler)
    model = init_model(;
        envmap = envmap(20),
        pr_killed = 0.075,
        seed = seed,
        scheduler = scheduler,
    )
    abm_video(
        videopath,
        model,
        agent_step!,
        model_step!;
        frames = 2880,
        framerate = 30,
        heatkwargs = heatkwargs,
        heatarray = model -> model.food,
        am = agent -> :circle,
    )
    return videopath
end
```
which means I'll need to dig a little deeper still.

**Comment by Libbum** [comment]
OK, yes I think I've identified the issue. Will port the rest of the problem over to JuliaDynamics/InteractiveDynamics.jl#54

File: src/simulations/collect.jl [hunk 1]
```diff
@@ -209,7 +209,7 @@ function init_agent_dataframe(model::ABM{S,A}, properties::AbstractArray) where
 end
 
 function single_agent_types!(types::Vector{Vector{T} where T}, model::ABM, properties::AbstractArray)
-    a = random_agent(model)
+    a = model[1]
```
[the remark this answers is not in the document]
Comment (Libbum) [reply src/simulations/collect.jl:212]:
```julia
julia> first(model.agents)
56 => BPH(56, (3, 4), Float16(0.6553), 256)
```
It returns the key value pair, so you want the agent.

[the remark this answers is not in the document]
Comment (Datseris) [reply src/simulations/collect.jl]: oh, didn't know that `.second` also achieved that, I would expect `[2]` at the end.

**Pull request merged**